### ISSUE TITLE
Flutter macOS desktop app freezes after switching applications using Mission Control.

### ISSUE THREAD

**Issue by chitraarasu** [flutter/flutter#155977]
### Steps to reproduce

1. Create a new Flutter project: flutter create bug.
2. Run the application and switch to another app. Then, reopen the Flutter app using Mission Control.
3. Try this multiple times—you will likely encounter the issue within one or two tries. Also, check out the video I’ve provided, where I’ve recreated the issue. Interestingly, when I maximize the app from the bottom icons, it works fine.

### Expected results

App should not freeze if app opens from Mission Control.

### Actual results

App becomes not responding when it opens from mission control. 

Same bug reported on windows but i have tested on windows it works fine. #74797

### Code sample

```
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // TRY THIS: Try running your application with "flutter run". You'll see
        // the application has a purple toolbar. Then, without quitting the app,
        // try changing the seedColor in the colorScheme below to Colors.green
        // and then invoke "hot reload" (save your changes or press the "hot
        // reload" button in a Flutter-supported IDE, or press "r" if you used
        // the command line to start the app).
        //
        // Notice that the counter didn't reset back to zero; the application
        // state is not lost during the reload. To reset the state, use hot
        // restart instead.
        //
        // This works for code too, not just values: Most code changes can be
        // tested with just a hot reload.
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // TRY THIS: Try changing the color here to a specific color (to
        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
        // change color while the other colors stay the same.
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          //
          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
          // action in the IDE, or press "p" in the console), to see the
          // wireframe for each widget.
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'You have pushed the button this many times:',
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headlineMedium,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/078588c9-a8d0-4fc7-99ce-968b01858d43

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
kirshi@Kirshis-MacBook-Pro ~ % flutter doctor -v
[✓] Flutter (Channel stable, 3.24.3, on macOS 15.0 24A335 darwin-x64, locale
    en-US)
    • Flutter version 3.24.3 on channel stable at
      /Users/kirshi/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (3 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[!] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/kirshi/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/to/macos-android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✗] Chrome - develop for the web (Cannot find Chrome executable at
    /Applications/Google Chrome.app/Contents/MacOS/Google Chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.11+0-17.0.11b1207.24-11852314)

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.96.0

[✓] Connected device (1 available)
    • macOS (desktop) • macos • darwin-x64 • macOS 15.0 24A335 darwin-x64

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 2 categories.
kirshi@Kirshis-MacBook-Pro ~ % 

```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @chitraarasu 
I tried the same on my mac but wasn't able to replicate the reported behavior. I tried almost 10 to 15 times to be sure.
I used trackpad to bring up or use mission control key by swiping with four fingers on it. Are you also using trackpad or mouse ?
Also, can you check if it occurs with native macOS app as well (ex: notes) ? 
I see that you are on latest macOS which I am not on, maybe that can be a factor as well.

**Comment by Tedsterh** [comment]
Heya, I am also experiencing the same issue, specifically when I have the app running in a second desktop and I swipe between, the macOS app freezes and when I try to hot reload from vscode the app crashes and I have to rebuild it.

I have recently upgraded my flutter to version 3.24.3, here is my flutter doctor output as well:
```
[✓] Flutter (Channel stable, 3.24.3, on macOS 14.6.1 23G93 darwin-arm64, locale en-GB)
    • Flutter version 3.24.3 on channel stable at /Users/edwardharcombe/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (3 weeks ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/edwardharcombe/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11609105)

[✓] VS Code (version 1.91.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.94.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.6.1 23G93 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.6.1 23G93 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome 129.0.6668.60

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

**Comment by chitraarasu** [comment]
Hi @darshankawar, 
1. I have tried on both mouse and trackpad. i am facing this issue on both. 
2. Other native application focus works fine.

**Comment by darshankawar** [comment]
Thanks for the update. I tried it again but was not able to replicate, but my colleague was able to replicate it once but not after that, so we can't be sure to provide a concrete pattern that leads to the reported behavior. 
Colleague tried on macOS 15, while [this](https://github.com/flutter/flutter/issues/155977#issuecomment-2385297644) observation was on macOS 14.6

<details>
<summary>flutter doctor -v (stable and master)</summary>

```console
[✓] Flutter (Channel stable, 3.24.3, on macOS 15.0 24A335 darwin-x64, locale en-VN)
    • Flutter version 3.24.3 on channel stable at /Users/huynq/Documents/GitHub/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2663184aa7 (11 days ago), 2024-09-11 16:27:48 -0500
    • Engine revision 36335019a8
    • Dart version 3.5.3
    • DevTools version 2.37.3

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio Koala.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] IntelliJ IDEA Ultimate Edition (version 2024.2.0.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 81.1.3
    • Dart plugin version 242.20629

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.96.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 15.0 24A335 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 128.0.6613.138

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

```console
[!] Flutter (Channel master, 3.26.0-1.0.pre.307, on macOS 15.0 24A335 darwin-x64, locale en-VN)
    • Flutter version 3.26.0-1.0.pre.307 on channel master at /Users/huynq/Documents/GitHub/flutter_master
    ! Warning: `flutter` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/huynq/Documents/GitHub/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/huynq/Documents/GitHub/flutter_master. Consider adding /Users/huynq/Documents/GitHub/flutter_master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 3dc1e81526 (20 minutes ago), 2024-10-01 19:17:06 -0700
    • Engine revision a5bc2e2708
    • Dart version 3.6.0 (build 3.6.0-309.0.dev)
    • DevTools version 2.40.0-dev.2
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
    • Android SDK at /Users/huynq/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • ANDROID_HOME = /Users/huynq/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 16.0)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 16A242d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2024.1)
    • Android Studio at /Applications/Android Studio Koala.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.11+0-17.0.11b1207.24-11852314)

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • android-studio-dir = /Applications/Android Studio.app/
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] IntelliJ IDEA Community Edition (version 2024.2.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 81.1.3
    • Dart plugin version 242.22855.32

[✓] VS Code (version 1.93.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.98.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-x64     • macOS 15.0 24A335 darwin-x64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 129.0.6668.70

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```
</details>

Since it seems to be intermittent, I will keep it open for team's tracking.

**Comment by codesession81** [comment]
Iam experiencing the same problem, running the latest OSX, flutter (stable channel) and XCode version.

When i switch between vscode and the app with mission controll the app freezes, hot reload or restart shows no effect.

I´ve tried debugging mode but get no errors or thrown exceptions.

It doesn`t matter if i run the default or a bigger project, in both cases the problem occours, so posting code seems to be useless.

Is there any workaround?

**Comment by jerefrer** [comment]
I don't know if that's exactly the same thing, but I started debugging my app as macOS target, which worked fine until I decided  to expand the app into its own space, then drag a browser with YouTube tutorial I was following on that space to have a split view. I think at first it worked, I could even move the vertical line to resize the spaces, but then I don't remember which actions I did and it just froze.

After that whenever I run the app it starts as full screen in its own space and is immediately frozen. I can't un-expand it back to the normal desktop nor do anything else.

Any idea what to do ?

Thanks !

**Comment by Princewil** [comment]
> Iam experiencing the same problem, running the latest OSX, flutter (stable channel) and XCode version.
> 
> When i switch between vscode and the app with mission controll the app freezes, hot reload or restart shows no effect.
> 
> I´ve tried debugging mode but get no errors or thrown exceptions.
> 
> It doesn`t matter if i run the default or a bigger project, in both cases the problem occours, so posting code seems to be useless.
> 
> Is there any workaround?

I am facing similar issue

**Comment by ahmeedev** [comment]
**I'm Facing this same issue**

**Comment by belizariogr** [comment]
Same issue... is there a workaround to fix it?

**Comment by Jishnu-p-ibo** [comment]
I am facing this issue

**Comment by ellie-ka-nam** [comment]
### Additional reproduction + detailed diagnostics (macOS render freeze after occlusion)

We are hitting this on a production macOS Flutter app and did fairly deep debugging. Sharing in case it helps narrow the root cause.

**Environment**
- Flutter 3.38.3 (stable), Engine revision 8bf2090718 (also reproduced on older 3.x)
- Reproduces on **Intel Macs** (low-spec/integrated-GPU machines) very reliably; rarely or never on faster Apple-Silicon machines
- App embeds a `WKWebView` (via `flutter_inappwebview`) on the main screen, but the issue is **not** webview-specific (see below)

**Symptom**
- Switch to another app (Cmd-Tab / Mission Control / another app's window covering ours) and switch back → the whole Flutter UI is frozen: no hover feedback, clicks do nothing, and any continuously-updating widgets stop updating.
- Deterministic ordering: **the 1st return is fine; it freezes from the 2nd return onward.**

**Key differentiators we found**
1. **Launch method matters.** Double-click / `open` (LaunchServices GUI launch) → freezes. Running the same binary directly from Terminal (`Foo.app/Contents/MacOS/Foo`) → **never freezes**, no matter how many times we switch away/back.
2. **Recovery is possible via the Dock.** When frozen, clicking the app's **Dock icon brings it back to life** (takes two clicks: the 1st doesn't bring the window forward, the 2nd does and rendering resumes). Cmd-Tab / Mission Control return does **not** recover it.

**What the freeze looks like in a `sample` (spindump-style)**
At the moment of freeze, **every thread is idle/parked** — there is no busy/spinning thread and nothing blocked on a lock:
- `main thread`: parked in `-[NSApplication run]` → `_DPSNextEvent` → `mach_msg2_trap` (normal idle event loop)
- `io.flutter.ui`: parked in `CFRunLoopRunSpecific` → `__CFRunLoopRun` → `mach_msg` (waiting for a vsync wakeup that never comes)
- `io.flutter.raster`: parked in `CFRunLoopRunSpecific` → `mach_msg` — **not** in `[CAMetalLayer nextDrawable]`, **not** in any semaphore wait
- all worker threads: `__psynch_cvwait`

So it is **not** a deadlock and **not** a `nextDrawable`/GPU stall — it looks exactly like **frame production / the display-link vsync simply stops being driven and never resumes** after the occlusion→visible transition. The Dart isolate is alive (we caught it mid-execution in some samples), it's just never ticked for a new frame.

**Things we tried that did NOT fix it (all confirmed active in the tested build via logging):**
- Disabling App Nap (`ProcessInfo.beginActivity(.userInitiatedAllowingIdleSystemSleep)`, also `NSAppSleepDisabled`) — no effect.
- Disabling the experimental merged platform/UI thread (`FLTEnableMergedPlatformUIThread=false`) — confirmed off (separate `io.flutter.ui` thread appears), still freezes.
- Dart-side `WidgetsBinding.instance.scheduleForcedFrame()` on resume — no effect (the display link is the gate; forcing a frame request doesn't restart it).
- Native window nudges on the `NSWindow.didChangeOcclusionStateNotification` "became visible" transition: `setNeedsDisplay` + a 1px window resize, `makeKeyAndOrderFront` + `orderFrontRegardless`, and even `orderOut(nil)` followed by `makeKeyAndOrderFront` (programmatically re-showing the window) — **none recover rendering**, even though a real user Dock-click (which is the same conceptual "re-show") does.

The fact that a genuine Dock-click recovers it but no programmatic `orderOut`/`orderFront`/`makeKeyAndOrderFront` does, suggests the recovery requires a real WindowServer-driven occlusion/visibility event that the engine observes to restart the display link — and the normal Cmd-Tab/Mission-Control return path is **not** delivering whatever signal the engine needs to resume frame production.

**Notes**
- `applicationDidBecomeActive` does **not** fire on these app-switch transitions for us (the app does not fully resign/become active on occlusion), so any workaround hooked there is dead code. `NSWindow.didChangeOcclusionStateNotification` does fire reliably.
- Happy to provide full `sample` outputs for the frozen state if useful.

**Comment by ellie-ka-nam** [comment]
**Update: root cause found, and a fix PR is up → #188772**

Following up on my earlier comment. I built a local engine with tracing in `FlutterEngine` (app-lifecycle) plus `FlutterVSyncWaiter` / `FlutterDisplayLink` (vsync) and reproduced the freeze on a low-spec Intel Mac. The trace showed the vsync/display-link layer is actually **healthy** — my earlier "the display link stops and never resumes" reading was a symptom, not the cause. The engine simply was never *asked* to render, because the framework had been told it was hidden.

**Root cause** is in `FlutterEngine`'s lifecycle handling:

```objc
- (void)handleWillBecomeActive:(NSNotification*)notification {
  _active = YES;
  if (!_visible) {
    [self setApplicationState:flutter::AppLifecycleState::kHidden];
  } else {
    [self setApplicationState:flutter::AppLifecycleState::kResumed];
  }
}
```

`_visible` is updated **only** by `handleDidChangeOcclusionState`. macOS does not reliably deliver an `NSApplicationDidChangeOcclusionState`(visible) notification on every occlusion→visible transition — returning to the app **on the same screen** (Cmd-Tab / Mission Control / clicking away and back) frequently delivers `visible=NO` when leaving but **never delivers `visible=YES` when returning**. So:

1. App occluded → `kHidden`, `_visible = NO` (frames disabled — correct).
2. App returns to the foreground → `applicationWillBecomeActive` fires, but `_visible` is still stale-`NO`, so `kHidden` is sent **again**.
3. Framework stays hidden: animations muted, `scheduleFrame` gated, no vsync requested → **frozen UI even though the window is on-screen and frontmost.**

The trace at the freezing return:

```
handleDidChangeOcclusionState visible=0  -> AppLifecycleState.hidden
handleWillBecomeActive   _visible(prev)=0   occlusionState&Visible=0   -> AppLifecycleState.hidden   (stuck)
(no further waitForVSync / onDisplayLink)
```

i.e. `willBecomeActive` *does* fire on the return, but `[[NSApplication sharedApplication] occlusionState]` still reads not-visible at that instant and the `visible` notification never arrives. This also explains the earlier observations: programmatic `orderFront`/`makeKeyAndOrderFront` didn't help (they don't update the lifecycle), while a real Dock-icon activation did (it eventually drives the occlusion state back to visible). Terminal-launched processes don't hit the aggressive occlusion behavior, so they never freeze.

**Fix** (in #188772): an app receiving `applicationWillBecomeActive` is frontmost and visible, so treat that as an authoritative "visible" signal and resume. `handleDidChangeOcclusionState` still drives `kHidden` when genuinely occluded, so the app can still go hidden.

```objc
- (void)handleWillBecomeActive:(NSNotification*)notification {
  _active = YES;
  _visible = YES;
  [self setApplicationState:flutter::AppLifecycleState::kResumed];
}
```

Verified on the Intel repro with a custom engine build: with this change the UI keeps rendering across every occlusion→return cycle. PR includes a regression test in `FlutterEngineTest.mm`.

**Workaround until it lands in a stable release:** when frozen, click the app's **Dock icon** to bring it back (rather than Cmd-Tab / Mission Control).

**Comment by cbracken** [comment]
If I'm understanding correctly, this was introduced in flutter/engine#40542, reverted in flutter/engine#42384, relanded in flutter/engine#42418. In the original code `_visible` is only ever updated by the occlusion notification and `handleWillBecomeActive` relies on that.

That code dates back to 2023; looks like this was reported about a year and a little after so that tracks. @ellie-ka-nam thanks for the excellent analysis and the fix.
